### ISSUE TITLE
S3 Example broken

### ISSUE THREAD

**Issue by reactormonk** [brendanhay/amazonka#466]
This line now breaks with a `ConnectionClosed`. It worked before with lts-9.X, doesn't anymore with nightly stackage.

From what I know, it's an issue with the connection being closed too fast in this part of the code. Instead, the `MonadResource` instance from outside should be used.
https://github.com/brendanhay/amazonka/blob/develop/examples/src/Example/S3.hs#L76

**Comment by AusCoder** [comment]
I ran into the same problem using amazonka-1.6.0. Yes, it would appear the connection is being closed too soon. Here is my best guess as to why: in `Network.AWS.Internal.HTTP`, the http call is made with
```
perform Env{..} x = liftIO $ catches (runResourceT go) handlers
```
My guess is that the runResourceT is closing the connection before we can actually use the conduit.

However, I think the problem has been fixed. This line was reverted recently, [see here](https://github.com/brendanhay/amazonka/blob/develop/amazonka/src/Network/AWS/Internal/HTTP.hs#L134). It appears to work with a local copy of master.

**Comment by ababkin** [comment]
can also confirm this is broken on 06/02 nightly stackage (1.6.0)

**Comment by blender** [comment]
Adding my voice to the others, broken on lts 11.13 .

**Comment by utdemir** [comment]
@brendanhay Is possible to have a new release, since this is already fixed on `develop`?

**Comment by naushadh** [comment]
For the future, something like https://github.com/brendanhay/amazonka/issues/477 can help prevent such regressions.

**Comment by jdreaver** [comment]
FYI it appears this is the PR that fixed the line @AusCoder is talking about https://github.com/brendanhay/amazonka/pull/464

**Comment by dfithian** [comment]
Would also LOVE a release

**Comment by ryota-ka** [comment]
Are there any workarounds this problem until the new version is released?
(Or I just have to wait for it?)

**Comment by blender** [comment]
@ryota-ka if you use stack you can use the latest master like so:
https://github.com/blender/Rome/blob/master/stack.yaml#L38-L46

**Comment by newhoggy** [comment]
@brendanhay can we have a new release please.  🙏

**Comment by dsturnbull** [comment]
Would really love a patch release 🙏

**Comment by MichaelXavier** [comment]
Relevant to other folks in this thread, see #484 for a status update. @brendanhay is going to be committing some time to work on this project in late August/early September. If anyone is a heavy user that wants to help with the maintenance burden of the project, release management, etc, perhaps get in touch.

**Comment by greghill** [comment]
@blender it turns out I was trying to read the conduit after already finishing `runResourceT` so I got the same error as this one even after applying your fix. I fixed that and now your configuration works.

The joys of software

**Comment by rblnguyen** [comment]
thank you @blender. The commit you pointed out work great. 
git: git@github.com:brendanhay/amazonka.git
 commit: c9d8a62118b2269fb54b1e89c4e92106e8243050
 subdirs:
   - amazonka
   - core
   - amazonka-s3

**Comment by neongreen** [comment]
A warning: I'm seeing significant slowdowns when talking to SQS, introduced by this commit: https://github.com/brendanhay/amazonka/commit/2688190f659e4d87fd6862b52b04be706b17d251. I'm investigating currently; maybe it would be prudent to avoid using the HEAD commit for now.

**Comment by neongreen** [comment]
https://github.com/brendanhay/amazonka/issues/490

**Comment by ianbamforth** [comment]
I also had the issue where the `HttpExceptionRequest ` was returned as above. Using the commit as advised by @blender got me past that (thanks!), but the file written to disk is always empty.

`downloadAndReadInFile :: AWS.Env -> S3Location -> IO (Either SomeException Text)
downloadAndReadInFile awsEnv loc@(S3Location bucket key) = AWS.runResourceT $ do
  downloadResult <- try $ AWS.runAWS awsEnv $ do
    res <- AWS.send (S3.getObject (S3.BucketName bucket) (S3.ObjectKey key))
    AWS.sinkBody (res ^. S3.gorsBody) (sinkSystemTempFile fileName)
  case downloadResult of
    Left ex -> pure $ Left ex
    Right fp -> liftIO $ mapLeft SomeException <$> readPlainText fp ext
  where
  ext = takeExtension fileName
  fileName = unpack $ s3FileName loc
`

Any idea why this could be happening?

**Comment by blender** [comment]
Have anyone tried if this is fixed in latest release 1.6.1 and can be closed?

**Comment by ysangkok** [comment]
There was a thread in the Discourse about preventing handles from leaking: https://discourse.haskell.org/t/bluefin-prevents-handles-leaking/10627

It seems that one could use a rank-2 type like in the example MangoIV posted. This would need to be added to the AWS monad, which would mean switching away from ResourceT.

EDIT: While the example is no longer broken, I don't think the fundamental footgun pointed out by @greghill has been fixed.
